File: src_py/prompt_guard/tool_args.py
"""Detect injection patterns in agent tool / function call arguments.

Unlike free-text prompt injection, tool args are usually structured
(`dict[str, Any]`) and the threats are concrete: SQL/NoSQL injection,
shell metacharacters, path traversal, SSRF to cloud metadata endpoints,
and code-interpreter primitives. This module walks the args recursively
and scans every string value with a dedicated rule set.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Iterable, List, Mapping, Pattern, Sequence, Tuple

from .config import DEFAULT_CONFIG, ScanConfig, Thresholds
from .types import Finding, category_for_label, severity_for_score


Rule = Tuple[str, Pattern[str], float]


TOOL_ARGS_RULES: Tuple[Rule, ...] = (
    (
        "sql_injection",
        re.compile(
            r"('\s*or\s*'?\d|\bunion\s+select\b|\bdrop\s+table\b|\b(?:or|and)\s+\d+\s*=\s*\d+\b|--(?:\s|$)|/\*.*?\*/|;\s*--)",
            re.IGNORECASE | re.DOTALL,
        ),
        0.7,
    ),
    (
        "nosql_operator_injection",
        re.compile(r'(?<![A-Za-z])\$(?:where|ne|gt|gte|lt|lte|regex|in|nin|exists|expr)\b'),
        0.6,
    ),
    (
        "shell_metacharacter",
        re.compile(r"(`[^`]+`|\$\([^)]+\)|\|\s*(?:bash|sh|zsh|nc|curl|wget|python|perl|ruby)\b|;\s*(?:rm|curl|wget|nc|bash|sh)\s)"),
        0.7,
    ),
    (
        "path_traversal",
        re.compile(r"(?:\.\./|\.\.\\){2,}|(?:^|/)etc/(?:passwd|shadow)\b|(?:^|\\)windows\\system32\b", re.IGNORECASE),
        0.7,
    ),
    (
        "ssrf_metadata_url",
        re.compile(
            r"\bhttps?://(?:169\.254\.169\.254|metadata\.google\.internal|metadata\.azure\.com|fd00:ec2::254)\b",
            re.IGNORECASE,
        ),
        0.85,
    ),
    (
        "ssrf_private_url",
        re.compile(
            r"\bhttps?://(?:127\.\d+\.\d+\.\d+|localhost|0\.0\.0\.0|10\.\d+\.\d+\.\d+|192\.168\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+)",
            re.IGNORECASE,
        ),
        0.6,
    ),
    (
        "code_interpreter_primitive",
        re.compile(
            r"\b(?:import\s+(?:os|subprocess|socket|requests)\b|__import__\s*\(|subprocess\.|os\.system\s*\(|eval\s*\(|exec\s*\(|compile\s*\()",
            re.IGNORECASE,
        ),
        0.7,
    ),
    (
        "windows_command_injection",
        re.compile(r"(?:^|[&|;])\s*(?:cmd|powershell|wmic|reg|net)\.?(?:exe)?\s+", re.IGNORECASE),
        0.65,
    ),
)


@dataclass(frozen=True)
class ToolArgsReport:
    args: Any
    tool_args_score: float
    tool_args_findings: List[Finding] = field(default_factory=list)
    thresholds: Thresholds = field(default_factory=Thresholds)

    @property
    def is_tool_args_injection(self) -> bool:
        return self.tool_args_score >= self.thresholds.tool_args

    def as_dict(self) -> dict:
        return {
            "is_tool_args_injection": self.is_tool_args_injection,
            "tool_args_score": self.tool_args_score,
            "tool_args_findings": [
                {
                    "kind": f.kind,
                    "label": f.label,
                    "evidence": f.evidence,
                    "path": _path_from_evidence_id(f),
                    "score": f.score,
                    "severity": f.severity,
                    "category": f.category,
                }
                for f in self.tool_args_findings
            ],
            "thresholds": {
                "tool_args": self.thresholds.tool_args,
            },
        }


def detect_tool_args(args: Any, config: ScanConfig = DEFAULT_CONFIG) -> Tuple[float, List[Finding]]:
    findings: List[Finding] = []
    for path, value in _iter_string_values(args, ""):
        for label, pattern, score in _configured_rules(config):
            for match in pattern.finditer(value):
                evidence = match.group(0).strip()
                # Encode the arg path into the evidence prefix to keep the
                # Finding shape compatible with the rest of the kit.
                findings.append(
                    Finding(
                        kind="tool_args",
                        label=label,
                        evidence=f"{path}: {evidence}" if path else evidence,
                        start=-1,
                        end=-1,
                        score=score,
                        severity=severity_for_score(score),
                        category=category_for_label(label, "tool_args"),
                    )
                )
    return _aggregate(findings), findings


def scan_tool_args(args: Any, config: ScanConfig = DEFAULT_CONFIG) -> ToolArgsReport:
    score, findings = detect_tool_args(args, config=config)
    return ToolArgsReport(
        args=args,
        tool_args_score=score,
        tool_args_findings=findings,
        thresholds=config.thresholds,
    )


def _configured_rules(config: ScanConfig) -> Iterable[Rule]:
    rules = list(TOOL_ARGS_RULES)
    for item in config.extra_rules:
        if item.kind != "tool_args":
            continue
        flags = 0
        if not item.case_sensitive:
            flags |= re.IGNORECASE
        if item.dotall:
            flags |= re.DOTALL
        rules.append((item.label, re.compile(item.pattern, flags), item.score))
    if config.enabled_rules is not None:
        builtin_labels = {label for label, _, _ in TOOL_ARGS_RULES}
        if config.enabled_rules & builtin_labels:
            rules = [r for r in rules if r[0] in config.enabled_rules or r[0] not in builtin_labels]
    if config.disabled_rules:
        rules = [r for r in rules if r[0] not in config.disabled_rules]
    return rules


def _iter_string_values(value: Any, path: str) -> Iterable[Tuple[str, str]]:
    if isinstance(value, str):
        yield path, value
    elif isinstance(value, Mapping):
        for key, sub in value.items():
            child = f"{path}.{key}" if path else str(key)
            yield from _iter_string_values(sub, child)
    elif isinstance(value, (list, tuple)) and not isinstance(value, (str, bytes)):
        for index, sub in enumerate(value):
            child = f"{path}[{index}]" if path else f"[{index}]"
            yield from _iter_string_values(sub, child)
    # ints/floats/bools/None: nothing to scan.


def _aggregate(findings: Sequence[Finding]) -> float:
    if not findings:
        return 0.0
    miss = 1.0
    for f in findings:
        miss *= max(0.0, 1.0 - f.score)
    score = 1.0 - miss
    if len(findings) >= 2:
        score = min(1.0, score + 0.1)
    return round(score, 3)


def _path_from_evidence_id(finding: Finding) -> str:
    # Evidence is "<path>: <snippet>"; the path is the prefix before the first ': '.
    if ": " in finding.evidence:
        return finding.evidence.split(": ", 1)[0]
    return ""

File: src_ts/tool_args.ts
/**
 * Detect injection patterns in agent tool / function call arguments.
 *
 * Unlike free-text prompt injection, tool args are usually structured and the
 * threats are concrete: SQL/NoSQL injection, shell metacharacters, path
 * traversal, SSRF to cloud metadata endpoints, code-interpreter primitives.
 */

import { normalizeConfig, type ScanConfig, type ScanConfigInput, type Thresholds } from "./config.js";
import { categoryForLabel, severityForScore, type Finding } from "./types.js";

type Rule = [label: string, pattern: RegExp, score: number];

const TOOL_ARGS_RULES: Rule[] = [
  [
    "sql_injection",
    /('\s*or\s*'?\d|\bunion\s+select\b|\bdrop\s+table\b|\b(?:or|and)\s+\d+\s*=\s*\d+\b|--(?:\s|$)|\/\*[\s\S]*?\*\/|;\s*--)/gi,
    0.7,
  ],
  ["nosql_operator_injection", /(?<![A-Za-z])\$(?:where|ne|gt|gte|lt|lte|regex|in|nin|exists|expr)\b/g, 0.6],
  [
    "shell_metacharacter",
    /(`[^`]+`|\$\([^)]+\)|\|\s*(?:bash|sh|zsh|nc|curl|wget|python|perl|ruby)\b|;\s*(?:rm|curl|wget|nc|bash|sh)\s)/g,
    0.7,
  ],
  [
    "path_traversal",
    /(?:\.\.\/|\.\.\\){2,}|(?:^|\/)etc\/(?:passwd|shadow)\b|(?:^|\\)windows\\system32\b/gi,
    0.7,
  ],
  [
    "ssrf_metadata_url",
    /\bhttps?:\/\/(?:169\.254\.169\.254|metadata\.google\.internal|metadata\.azure\.com|fd00:ec2::254)\b/gi,
    0.85,
  ],
  [
    "ssrf_private_url",
    /\bhttps?:\/\/(?:127\.\d+\.\d+\.\d+|localhost|0\.0\.0\.0|10\.\d+\.\d+\.\d+|192\.168\.\d+\.\d+|172\.(?:1[6-9]|2\d|3[01])\.\d+\.\d+)/gi,
    0.6,
  ],
  [
    "code_interpreter_primitive",
    /\b(?:import\s+(?:os|subprocess|socket|requests)\b|__import__\s*\(|subprocess\.|os\.system\s*\(|eval\s*\(|exec\s*\(|compile\s*\()/gi,
    0.7,
  ],
  ["windows_command_injection", /(?:^|[&|;])\s*(?:cmd|powershell|wmic|reg|net)\.?(?:exe)?\s+/gi, 0.65],
];

export interface ToolArgsReport {
  args: unknown;
  toolArgsScore: number;
  isToolArgsInjection: boolean;
  toolArgsFindings: Finding[];
  thresholds: Thresholds;
}

export function detectToolArgs(args: unknown, config: ScanConfigInput = {}): [number, Finding[]] {
  const normalized = normalizeConfig(config);
  const findings: Finding[] = [];
  const rules = configuredRules(normalized);
  for (const { path, value } of iterStringValues(args, "")) {
    for (const [label, pattern, score] of rules) {
      pattern.lastIndex = 0;
      for (const match of value.matchAll(pattern)) {
        const evidence = match[0].trim();
        findings.push({
          kind: "tool_args",
          label,
          evidence: path ? `${path}: ${evidence}` : evidence,
          start: -1,
          end: -1,
          score,
          severity: severityForScore(score),
          category: categoryForLabel(label, "tool_args"),
        });
      }
    }
  }
  return [aggregate(findings), findings];
}

export function scanToolArgs(args: unknown, config: ScanConfigInput = {}): ToolArgsReport {
  const normalized = normalizeConfig(config);
  const [score, findings] = detectToolArgs(args, normalized);
  return {
    args,
    toolArgsScore: score,
    isToolArgsInjection: score >= normalized.thresholds.toolArgs,
    toolArgsFindings: findings,
    thresholds: normalized.thresholds,
  };
}

function configuredRules(config: ScanConfig): Rule[] {
  let rules: Rule[] = [...TOOL_ARGS_RULES];
  for (const item of config.extraRules) {
    if (item.kind !== "tool_args") continue;
    const flags = "g" + (item.caseSensitive ? "" : "i") + (item.dotall ?? true ? "s" : "");
    rules.push([item.label, new RegExp(item.pattern, flags), item.score]);
  }
  const enabledRules = config.enabledRules ? new Set(config.enabledRules) : undefined;
  if (enabledRules) {
    const builtin = new Set(TOOL_ARGS_RULES.map(([label]) => label));
    if ([...enabledRules].some((label) => builtin.has(label))) {
      rules = rules.filter(([label]) => enabledRules.has(label) || !builtin.has(label));
    }
  }
  const disabledRules = new Set(config.disabledRules);
  if (disabledRules.size) rules = rules.filter(([label]) => !disabledRules.has(label));
  return rules;
}

function* iterStringValues(value: unknown, path: string): Generator<{ path: string; value: string }> {
  if (typeof value === "string") {
    yield { path, value };
    return;
  }
  if (Array.isArray(value)) {
    for (let i = 0; i < value.length; i += 1) {
      yield* iterStringValues(value[i], path ? `${path}[${i}]` : `[${i}]`);
    }
    return;
  }
  if (value !== null && typeof value === "object") {
    for (const [key, sub] of Object.entries(value as Record<string, unknown>)) {
      yield* iterStringValues(sub, path ? `${path}.${key}` : key);
    }
  }
}

function aggregate(findings: Finding[]): number {
  if (!findings.length) return 0;
  let miss = 1;
  for (const f of findings) miss *= Math.max(0, 1 - f.score);
  let score = 1 - miss;
  if (findings.length >= 2) score = Math.min(1, score + 0.1);
  return Math.round(score * 1000) / 1000;
}
